### ISSUE TITLE
Choice-frontier DAgger round (3 seeds): pre-registered separation from the exact-eps-0.75 rung on #135 panel + held-out P2 (GPU)

### ISSUE THREAD

**Issue by Sino-Huang** [Sino-Huang/multimodality_on_planning#140]
# Runbook: one DAgger round for the choice-frontier adapters (3 seeds) — pre-registered separation from the exact-eps-0.75 rung (GPU)

> The author is away and will not answer. Never ask questions or wait for replies. Use the **blocked exit** (rule 9) only for failures this ticket does not cover.

## 0. Why (context only)

#138: the 3-seed #136 adapters beat uniform choice (D3 = +0.285 [+0.140, +0.433], POSITIVE) but are **NOT_SEPARATED** from the exact-eps-0.75 rung (S = −0.041 [−0.187, +0.102]).

The adapters were trained purely by imitation on teacher-visited states (exact-reference episodes). At evaluation they drift into frontiers the teacher never produced — the classic compounding-error problem of behaviour cloning. One round of DAgger (Ross et al., 2011) targets exactly this:
1. roll out each adapter on training tasks;
2. label every visited decision with the teacher's choice;
3. continue training on the aggregate.

The reviewer's bar for a stronger rating is a learned result separated from the eps-0.75 rung by an interval.

## 1. Rules

1. Prefix Python with `source ~/cd_vlaplan &&`. Repo root: `/data/scratch/projects/punim0478/sukaih/multimodality_on_planning`. Always use `gh ... --repo Sino-Huang/multimodality_on_planning`.
2. Never read `data/bfws_phase_v1/fresh-test-manifest.jsonl` or `data/curriculum_pddl/*/dev/`.
3. **Frozen and read-only:**
   - existing `scripts/*choice_frontier*.py`
   - `outputs/choice-frontier/{v1,o4,v2,v3,v4}/`
   - `configs/experiments/choice-frontier{,-v2,-v3,-v4}/`

   Copy instead of editing. (#139 may still be writing under `outputs/choice-frontier/v4/panels/`; never touch it.)
4. **Rollout tasks come only from the #136 training tasks** (`configs/experiments/choice-frontier-v3/training-tasks.json`, rows with `excluded: false`), walked in the #136 frozen order. Assert that no rollout task's `problem_sha256` matches:
   - any #135 candidate (`configs/experiments/choice-frontier-v2/candidates.json`);
   - any #139 candidate (`configs/experiments/choice-frontier-v4/candidates.json`).
5. **GPU:**
   - New ledger `outputs/choice-frontier/v5/budget.json`, schedule `docs/experiments/choice-frontier/schedule-v5.json`: cap **40 GPU-h**, allocation `choice_frontier` 40, ports **18830-18833**, cutoffs ≥ 2026-10-20.
   - All GPU work goes through `scripts/run_expanded_study.py launch`, with a distinct MASTER_PORT per concurrent job and at most one job per GPU.
   - **Cross-ledger safety:** while `outputs/choice-frontier/v4/budget.json` has any `reserved` or `running` attempt, launch only on a GPU that attempt does not hold (currently GPU 1), and check `nvidia-smi` memory < 1 GB on that GPU first.
6. **Freeze order:**
   - Commit and push the protocol **before any rollout**.
   - Commit and push the on-policy corpus membership **before any training**.
   - Commit and push code and job files before their first launch.
7. `git add` only your own files; `git pull --rebase` before every push (another agent may be committing #139 files).
8. Skip formatters and project-wide test suites; run only the named pytest/ruff commands.
9. **Blocked exit:**
   - commit and push what is done;
   - write `docs/experiments/choice-frontier/issue-140-closeout.md` with a `## Blocked` section;
   - comment `## #140 blocked …` on the issue;
   - leave the issue OPEN.

## 2. Step 1 — protocol (commit before any rollout)

Write `docs/experiments/choice-frontier/issue-140-protocol.md` and `configs/experiments/choice-frontier-v5/protocol.json` with exactly these decisions.

**Base policies.** The six adapters (algorithm × seed):
- seed 17: `outputs/choice-frontier/v3/training/<alg>/seed-17/final/`
- seeds 29 and 71: `outputs/choice-frontier/v4/seeds/training/<alg>/seed-<s>/final/`

**Rollout collection (per algorithm × seed; GPU).**
- Roll out adapter (alg, s) under the frozen choice-frontier contract on the #136 training tasks, in the #136 frozen walk order. Use greedy decoding, inference seed 17, and the #136 visual observation contract.
- Decision cap is 2 × R_task, where R_task is the uncapped exact-reference expansion count of that task under alg (from the #136 derivation).
- At every on-policy decision whose menu has ≥ 2 states, record a sample:
  - the menu and observation exactly as the model saw it;
  - the **teacher label** = the menu label of the exact heap head of the *current* frontier (`controller.frontier_head_state_id()`, minimum (priority, generation serial));
  - whether the adapter agreed with the teacher.
- The adapter's own choice drives the rollout, not the teacher's.
- Take at most **32** records per task episode (the first 32 in episode order), continuing until **1024** on-policy records per (alg, s).
- If the training tasks run out first, use every record, provided there are ≥ 512; otherwise take the blocked exit.
- Episodes are logged and independently replayed, like evaluation episodes.

**Aggregated training set (per alg, s).**
- The 1024 on-policy records, plus 1024 records sampled without replacement from the frozen #136 corpus for alg using `random.Random(f"replay:{alg}:{s}")`.
- Apply the #136 menu-order augmentation ×2 (`aug:{record_id}:0/1`; the target is the teacher state's label after permutation): 4096 samples.
- Sample order: `random.Random(f"order:dagger:{alg}:{s}")`.

**Training.**
- Continue training the base adapter (alg, s): load its LoRA weights and keep training them.
- 1 epoch, global batch 32, microbatch 1, lr 1e-4 cosine, warmup 0.03, bf16. LoRA r/alpha/dropout/targets unchanged from #136. Training seed s. `max_seconds_per_cell` 18000.
- `audit-train` checks: loaded-from path, r=64, alpha=128, steps = 128, seed, and sample count.

**Smoke gate.** The #132 smoke definition on the seed-17 DAgger adapters. A FAIL means verdict SMOKE_FAIL.

**Evaluation.**
- DAgger adapters (6 cells) on:
  - the frozen #135 panel (`configs/experiments/choice-frontier-v2/membership.json`);
  - the frozen #139 P2 panel (`configs/experiments/choice-frontier-v4/membership-p2.json`).
- Greedy decoding, inference seed 17, cap 2R_t (the #135/#139 exact expansions). Native views reuse `outputs/choice-frontier/v2/reference-views.json` and the #139 P2 view report, read-only; live renders go under v5.
- Every episode is replayed at finalize.
- **Reused, not rerun:**
  - pre-DAgger learned episodes: #135 panel from #136/#138; P2 from #139;
  - controls and the exact-eps ladder: #135 zoo and the #139 P2 zoo.
- If #139's P2 learned/zoo evidence is not complete when you reach analysis, wait for #139 to close (poll `gh issue view 139 ... --json state` with ~10 min sleeps).

**Primary endpoint (pre-registered).** Pooled 23-task set = #135 panel (12) ∪ P2 (11), clustered by task.
- S_pool = M1(DAgger, 3-seed mean) − M1(exact-eps-0.75, averaged over its 5 seeds).
- Paired task-cluster bootstrap: `random.Random(133)`, 10,000 draws, seeds averaged inside each draw, 95% percentile interval.
- Verdict:
  - `SEPARATED_ABOVE` if lo > 0;
  - `SEPARATED_BELOW` if hi < 0;
  - otherwise `NOT_SEPARATED`.

**Co-primary.**
- Δ_pool = M1(DAgger) − M1(pre-DAgger), paired by (task, algorithm, seed), 3-seed mean, same bootstrap.
- Verdict:
  - `IMPROVED` if lo > 0;
  - `EQUIVALENT` if −0.05 < lo and hi < 0.05;
  - `WORSE` if hi < 0;
  - otherwise `INCONCLUSIVE`.

**Secondary (descriptive).**
- Per-panel S, Δ, and D3 = M1(DAgger) − M1(random_valid) under the #138 rule.
- Per-seed M1 with CI.
- Teacher agreement − chance.
- Last-label rate.
- On-policy teacher-agreement rate during collection.
- Ladder position per panel.

**Budget estimate** (write it in):
- collection ≈ 6 × 1024 calls × ~6 s ≈ 10 GPU-h
- training ≈ 6 × 4096 × 2.75 s ≈ 19 GPU-h
- evaluation ≈ 12 × 0.5 ≈ 6 GPU-h
- total ≈ 35, cap 40

**Budget fallback (pre-registered).** If the scheduler would exceed 40, drop evaluation cells in the order s71, then s29, and report which cells were dropped.

Commit, push, and post the hash on the issue.

## 3. Step 2 — code and jobs

- Create `scripts/run_choice_frontier_v5_dagger.py`. Import the frozen machinery (v3 corpus/views/training, v4 panels evaluation) instead of copying large files.
- Stages:
  - `collect --algorithm --seed` (GPU)
  - `audit-collect` (replay rollouts; check teacher labels by recomputing the heap head)
  - `build-corpus` (writes `configs/experiments/choice-frontier-v5/membership.json` with record ids, the sample count and `membership_sha256`)
  - `train --algorithm --seed`
  - `audit-train`
  - `smoke` / `audit-smoke`
  - `evaluate-worker --panel {v2,p2} --algorithm --seed`
  - `audit-evaluate-worker`
  - `finalize`
- Completion hooks locate the attempt dir from the terminal path's parent (the #136/#138 fix).
- Job files go in `configs/experiments/choice-frontier-v5/`.

## 4. Steps 3-6 — run

1. Init the v5 ledger. Launch collection jobs, two at a time once GPU 0 is free (honour rule 5 until then).
2. Run audit-collect, then build-corpus, then commit and push the membership.
3. Train the 6 cells, two at a time.
4. Run smoke, then evaluate the 12 cells, then finalize (0 missing, 0 mismatch).
5. Relaunch infrastructure failures at most 2 times; never change the recipe.

## 5. Step 7 — analysis

- Write `scripts/analyze_choice_frontier_v5_dagger.py`, importing M1–M4 and the bootstrap from `scripts/analyze_choice_frontier_v2.py` / `_v4_seeds.py`.
- Output `outputs/choice-frontier/v5/metrics/analysis.json` with blocks `primary` {S_pool, ci95, per_seed, verdict}, `co_primary` {delta_pool, ci95, verdict}, `per_panel`, `per_seed`, `collection` and `ladder_position`.

## 6. Step 8 — tests, closeout

- Add `tests/planning_benchmark/test_choice_frontier_v5_dagger.py`. It must check:
  - the teacher label equals the current frontier heap head for a rollout that deviates from the teacher;
  - the replay sampler is deterministic;
  - the pooled verdict rule at its boundaries.
- Run pytest on that file and ruff on the new scripts.
- Write `docs/experiments/choice-frontier/issue-140-closeout.md` with the #136 closeout sections: verdicts first, evidence index with the first 16 hex of each sha256, and limitations (one DAgger round, privileged teacher, two panels).
- Post the closing comment and close:

```
## #140 closeout
Primary (pooled 23 tasks): <SEPARATED_ABOVE|NOT_SEPARATED|SEPARATED_BELOW|SMOKE_FAIL>, S_pool = <x> [<lo>, <hi>]
Co-primary DAgger vs pre-DAgger: <IMPROVED|EQUIVALENT|WORSE|INCONCLUSIVE>, Δ_pool = <x> [<lo>, <hi>]
Per panel: #135 S <x> [..], D3 <x> [..]; P2 S <x> [..], D3 <x> [..]
Keys for #131 (outputs/choice-frontier/v5/metrics/analysis.json): primary.*, co_primary.*, per_panel.*, per_seed.*, collection
GPU: <x>/40 GPU-h (ledger outputs/choice-frontier/v5/budget.json)
Commits: protocol <sha>, corpus membership <sha>, evidence <sha>
```

**Comment by Sino-Huang** [comment]
## #140 protocol frozen
Protocol committed and pushed before any rollout: 43d9152 (docs/experiments/choice-frontier/issue-140-protocol.md, configs/experiments/choice-frontier-v5/protocol.json, docs/experiments/choice-frontier/schedule-v5.json).
Recorded decisions: rollout tasks skip any #136 training task whose problem_sha256 matches a #135 or #139 candidate row (incl. the 267 rejected #139 overlap rows) or that lacks frozen #136 train views (177 tasks remain); observation overflow stops an episode without a record; train_visual gains an optional init_adapter for continue-training.

**Comment by Sino-Huang** [comment]
## #140 amendment A1 (orchestrator, on the author's behalf; decided before any DAgger training or evaluation outcome)

**Hard wall-clock deadline: 14:21 UTC today (3 h from 11:21 UTC).** After that, the author stops all jobs to start writing. The full 3-seed plan (about 12 h) cannot fit, so the pre-registered design is reduced as follows. Collection results for seed 17 exist, but no DAgger training or evaluation outcome exists yet.

1. **Seed 17 only (2 cells).** Cancel the running seed-29 collection jobs now and record them as cancelled, with their hours charged. Seed 71 is not run.
2. **Aggregated training set per algorithm:**
   - the **first 512** on-policy seed-17 records, in the frozen collection order;
   - plus 512 replay records, drawn with the same `Random("replay:{alg}:17")` sampler (first 512 draws);
   - then the #136 ×2 menu augmentation, giving **2048 samples** (64 steps at global batch 32).
   - Recipe otherwise unchanged. audit-train checks 64 steps and 2048 samples.
3. **Evaluation:** DAgger s17 on the #135 panel first, then on P2 (one algorithm per GPU). Smoke gate unchanged.
4. **Primary endpoint (same rule, seed 17 only):**
   - S_pool = M1(DAgger s17) − M1(eps-0.75) on the pooled #135 ∪ P2 set;
   - co-primary: Δ_pool = M1(DAgger s17) − M1(pre-DAgger s17), paired by task and algorithm;
   - same bootstrap and verdict thresholds.
5. **Deadline fallback:** if the P2 evaluation cannot finish by about 13:55 UTC, stop it at that time. Report the #135-panel-only S and Δ as **descriptive**, and state that the primary endpoint is not evaluable.
6. **Closeout, closing comment and closing the issue** must be done by 14:15 UTC in every case, even if they only record partial evidence.

Commit this into issue-140-protocol.md as 'Amendment A1', and push it before launching any training job.

**Comment by Sino-Huang** [comment]
## #140 closeout
Primary (pooled 23 tasks, seed 17 under Amendment A1): NOT_SEPARATED, S_pool = +0.115 [-0.026, +0.258]
Co-primary DAgger vs pre-DAgger: INCONCLUSIVE, Δ_pool = -0.030 [-0.114, +0.057]
Per panel: #135 S -0.034 [-0.197, +0.135], D3 +0.292 [+0.133, +0.466]; P2 S +0.278 [+0.082, +0.473], D3 +0.467 [+0.269, +0.670]
Keys for #131 (outputs/choice-frontier/v5/metrics/analysis.json): primary.*, co_primary.*, per_panel.*, per_seed.*, collection
GPU: 9.06/40 GPU-h (ledger outputs/choice-frontier/v5/budget.json)
Commits: protocol 43d9152 (A1 c22a926), corpus membership c22a926, evidence 3996600
Note: one DAgger round (512 on-policy + 512 replay, 64 steps, seed 17) does not measurably change choice quality; the held-out P2 separation above eps-0.75 reproduces, the pooled one does not.
